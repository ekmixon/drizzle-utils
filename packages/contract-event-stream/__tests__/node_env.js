/**
 * @jest-environment node
 */
const Web3 = require("web3");
const createNewBlock$ = require("@drizzle-utils/new-block-stream");
const { take, finalize, tap, toArray } = require("rxjs/operators");
const initTestChain = require("@drizzle-utils/test-chain");

const createContractEvent$ = require("../index");
const { eventMatcher } = require("./utils/propertyMatchers");

jest.setTimeout(20000);

describe("contract-event-stream tests in node environment", () => {
  let provider;
  let web3;
  let accounts;
  let contractInstance;
  let artifact;

  beforeAll(async () => {
    const testChain = await initTestChain({
      contract: {
        dirname: __dirname,
        filename: "SimpleStorageWithEvents.sol",
        contractName: "SimpleStorageWithEvents",
      },
    });

    ({ provider, web3, accounts, contractInstance } = testChain);

    artifact = {
      ...testChain.contractArtifact,
      // truffle-decoder needs this in artifact
      networks: {
        "4447": {
          address: contractInstance._address,
        },
      },
    };
  });

  afterAll(async () => {
    provider.close();
  });

  test("createContractEvent$ function exists", () => {
    expect(createContractEvent$).toBeDefined();
  });

  test("createContractEvent$ throws errors when required options fields not found", () => {
    const { _address: address } = contractInstance;
    const { abi } = artifact;

    expect(() => createContractEvent$()).toThrow();

    expect(() => createContractEvent$({ abi, address })).toThrow(
      new Error("The options object with web3 is required"),
    );

    expect(() => createContractEvent$({ web3, address })).toThrow(
      new Error("The options object with contract abi is required"),
    );

    expect(() => createContractEvent$({ web3, abi })).toThrow(
      new Error("The options object with contract address is required"),
    );

    const web3Http = new Web3("http://127.0.0.1:9545"); // HttpProvider
    expect(() =>
      createContractEvent$({ web3: web3Http, abi, address }),
    ).toThrow(
      new Error("Must provide newBlock$ when using http provider with web3"),
    );
  });

  test("fromPolling can track events emitted by send method", async done => {
    const newBlock$ = createNewBlock$({
      web3,
      pollingInterval: 200,
    });

    const { _address: address } = contractInstance;
    const { abi } = artifact;
    const event$ = createContractEvent$({ web3, abi, address, newBlock$ });

    // tap observable to make sure it emitted a "0" and then a "5"
    event$
      .pipe(
        take(2),
        toArray(),
        tap(vals => {
          expect(vals[0]).toMatchSnapshot(eventMatcher);
          expect(vals[1]).toMatchSnapshot(eventMatcher);
          expect(vals[0].returnValues.storedData).toBe("0");
          expect(vals[1].returnValues.storedData).toBe("5");
        }),
        finalize(() => {
          expect.assertions(4);
          done();
        }),
      )
      .subscribe();

    await contractInstance.methods.set(0).send({ from: accounts[0] });
    await contractInstance.methods.set(5).send({ from: accounts[0] });
  });

  test("fromSubscribe can track events emitted by send method", async done => {
    // Change constructor so we can test Websocket route
    // Note that ganache provider.constructor.name is Provider
    function WebsocketProvider() {}
    const web3Ws = new Web3(provider);
    web3Ws.currentProvider.constructor = WebsocketProvider;

    const { _address: address } = contractInstance;
    const { abi } = artifact;
    const event$ = createContractEvent$({ web3: web3Ws, abi, address });

    // tap observable to make sure it emitted a "0" and then a "5"
    event$
      .pipe(
        take(2),
        toArray(),
        tap(vals => {
          expect(vals[0]).toMatchSnapshot(eventMatcher);
          expect(vals[1]).toMatchSnapshot(eventMatcher);
          expect(vals[0].returnValues.storedData).toBe("0");
          expect(vals[1].returnValues.storedData).toBe("5");
        }),
        finalize(() => {
          expect.assertions(4);
          done();
        }),
      )
      .subscribe();

    await contractInstance.methods.set(0).send({ from: accounts[0] });
    await contractInstance.methods.set(5).send({ from: accounts[0] });
  });
});
