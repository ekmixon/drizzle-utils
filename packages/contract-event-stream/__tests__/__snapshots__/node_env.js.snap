// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contract-event-stream tests in node environment can track events emitted by send method 1`] = `
Object {
  "address": Anything,
  "blockHash": Anything,
  "blockNumber": Anything,
  "event": "Update",
  "id": Anything,
  "logIndex": Anything,
  "raw": Object {
    "data": Anything,
    "topics": Array [
      Anything,
    ],
  },
  "returnValues": Object {
    "0": BigNumber {
      "_hex": "0x00",
    },
    "storedData": BigNumber {
      "_hex": "0x00",
    },
  },
  "signature": Anything,
  "transactionHash": Anything,
  "transactionIndex": 0,
  "type": "mined",
}
`;

exports[`contract-event-stream tests in node environment can track events emitted by send method 2`] = `
Object {
  "address": Anything,
  "blockHash": Anything,
  "blockNumber": Anything,
  "event": "Update",
  "id": Anything,
  "logIndex": Anything,
  "raw": Object {
    "data": Anything,
    "topics": Array [
      Anything,
    ],
  },
  "returnValues": Object {
    "0": BigNumber {
      "_hex": "0x05",
    },
    "storedData": BigNumber {
      "_hex": "0x05",
    },
  },
  "signature": Anything,
  "transactionHash": Anything,
  "transactionIndex": 0,
  "type": "mined",
}
`;
